main/mpu9250.go

```diff
@@ -0,0 +1,182 @@
+
+	setSetting(0x6A, 0x20) // I2C Master mode.
+	setSetting(0x24, 0x0D) // I2C configuration multi-master, 400KHz.
+
+	// AK8963 init.
+
+	setSetting(0x25, 0x0C) // Set the I2C slave addres of AK8963 and set for write.
+	setSetting(0x26, 0x0B) // I2C slave 0 register address from where to begin data transfer.
+	setSetting(0x63, 0x01) // Reset AK8963.
+	setSetting(0x27, 0x81) // Enable I2C and set 1 byte.
+
+	setSetting(0x25, 0x0C) // Set the I2C slave addres of AK8963 and set for write.
+	setSetting(0x26, 0x0A) // I2C slave 0 register address from where to begin data transfer.
+	setSetting(0x63, 0x16) // Register value to 100Hz continuous measurement in 16bit.
+	setSetting(0x27, 0x81) // Enable I2C and set 1 byte.
+
+	// Accelerometer and gyro init.
+
+	setSetting(0x19, 0x00) // Set Gyro 1000 Hz sample rate. rate = gyroscope output rate/(1 + value)
+	setSetting(0x1A, 0x03) // Set low pass filter to 92 Hz.
+	setSetting(0x1B, 0x00) // Set gyro sensitivity to 250dps.
+	setSetting(0x1C, 0x00) // Set accelerometer scale to +/- 2G.
+	setSetting(0x1D, 0x02) // Set Accel 1000 Hz sample rate.
+
+	if !checkMagConnection() {
+		log.Printf("magnetometer is offline.\n")
+		return
+	}
+
+	go readRawData()
+	go calculateAttitude()
+}
+
+func readRawData() {
+	timer := time.NewTicker(2 * time.Millisecond)
+
+	for {
+		<-timer.C
+		// Get accelerometer data.
+		x_acc, err := i2cbus.ReadWordFromReg(0x68, 0x3B)
+		chkErr(err)
+		y_acc, err := i2cbus.ReadWordFromReg(0x68, 0x3D)
+		chkErr(err)
+		z_acc, err := i2cbus.ReadWordFromReg(0x68, 0x3F)
+
+		// currently manually setting resolution
+		x_acc_f := float64(int16(x_acc)) * 0.00006103515625
+		y_acc_f := float64(int16(y_acc)) * 0.00006103515625
+		z_acc_f := float64(int16(z_acc)) * 0.00006103515625
+
+		// Get gyro data.
+		x_gyro, err := i2cbus.ReadWordFromReg(0x68, 0x43)
+		chkErr(err)
+		y_gyro, err := i2cbus.ReadWordFromReg(0x68, 0x45)
+		chkErr(err)
+		z_gyro, err := i2cbus.ReadWordFromReg(0x68, 0x47)
+
+		x_gyro_f := float64(int16(x_gyro)) * 0.00006103515625
+		y_gyro_f := float64(int16(y_gyro)) * 0.00006103515625
+		z_gyro_f := float64(int16(z_gyro)) * 0.00006103515625
+
+		// Get magnetometer data.
+		setSetting(0x25, 0x0C|0x80) // Set the I2C slave addres of AK8963 and set for read.
+		setSetting(0x26, 0x03)      // I2C slave 0 register address from where to begin data transfer.
+		setSetting(0x27, 0x87)      // Read 7 bytes from the magnetometer (HX+HY+HZ+ST2).
+		x_mag, err := i2cbus.ReadWordFromReg(0x68, 0x49)
+		chkErr(err)
+		y_mag, err := i2cbus.ReadWordFromReg(0x68, 0x4B)
+		chkErr(err)
+		z_mag, err := i2cbus.ReadWordFromReg(0x68, 0x4D)
+
+		st2, err := i2cbus.ReadByteFromReg(0x68, 0x4F) // ST2 register. Unlatch measurement data for next sample.
+		chkErr(err)
+
+		if st2&0x08 != 0 { // Measurement overflow. HOFL.
+			fmt.Printf("mag: measurement overflow\n")
+			continue // Don't use measurement.
+		}
+
+		x_mag_f := float64(int16(x_mag))*1.28785103785104*magXcal - 470.0
```

Comment:
Two changes are needed here:

1. The AK8963 magnetic sensor is physically located on a separate die from the gyro/accelerometer within the MPU9250 package. Orientation relative to the gyro die is flipped and rotated 90 degrees, such that mag_x = gyro_y, mag_y = gyro_x, and mag_z = -gyro_z.

   ![image](https://cloud.githubusercontent.com/assets/14127245/16768627/ab0b23c8-480a-11e6-99e5-5b1db75ac886.png)


2. You're hard coding hard iron corrections to the raw readings... why? These will be different for every installation, and the Madgwick method compensates for them anyway.

Suggested changes:

```
x_mag_f := float64(int16(y_mag)) * 1.28785103785104 * magXcal
y_mag_f := float64(int16(x_mag)) * 1.28785103785104 * magYcal
z_mag_f := float64(int16(-z_mag)) * 1.28785103785104 * magZcal
```